### PR TITLE
Add prometheus metrics sample

### LINKED ISSUE #833: Add prometheus library to vcpkg
Add this lib to the list of packages in vcpkg https://github.com/jupp0r/prometheus-cpp

### LINKED ISSUE #834: Metrics endpoint
Add flag `--metrics_port <port number>` which opens a port for requesting metrics of Iroha

### PR DESCRIPTION
**Pull request by kuvaldini**
Closes #833, closes #834

There is a config-file key `"metrics"` which represents
address:port in one line. Values could be addr:port, port,
or :port. I.e "127.0.0.1:8080", "9090", or ":1234".
See file `examples/config.sample`. Wrong value disables
Prometheus metrics.

There are two command line options to override config:
`--metrics_port` http port to listen on, default empty - disabled
`--metrics_addr` interface address to listen on, default 127.0.0.1

**In order to build must install two new packages with vcpkg**
```
/path/to/vcpkg install prometheus-cpp civetweb
```

### Usage Examples or Tests *[optional]*

```
>>> ./irohad --config example/config.sample --genesis_block example/genesis.block --keypair_name example/node0 --drop_state
[2021-03-08 19:01:36.438029000][I][Init]: Irohad version: 1.2.0-20-gc4d94e92
[2021-03-08 19:01:36.438744000][I][Init]: config initialized
*** 1615222896.439080000            0 123145526603776 mg_start2:19081: [listening_ports] -> [127.0.0.1:5555]
*** 1615222896.439110000        30000 123145526603776 mg_start2:19081: [num_threads] -> [2]
*** 1615222896.439372000       262000 123145526833152 consume_socket:18210: going idle
*** 1615222896.439400000        28000 123145526947840 consume_socket:18210: going idle
[2021-03-08 19:01:36.440390000][W][Init]: Using deprecated database connection string!
[2021-03-08 19:01:36.441339000][I][Irohad]: created
[2021-03-08 19:01:36.441702000][I][Irohad]: [Init] => pending transactions storage
[2021-03-08 19:01:36.810064000][I][Irohad]: [Init] => storage
[2021-03-08 19:01:36.810195000][W][Init]: Passed genesis block will be ignored without --overwrite_ledger flag. Restoring existing state.
[2021-03-08 19:01:36.823401000][I][Irohad/Storage/Storage/PostgresSettingQuery]: Kept value for MaxDescriptionSize: 64
[2021-03-08 19:01:36.823451000][I][Irohad]: [Init] => settings
[2021-03-08 19:01:36.823488000][I][Irohad]: [Init] => validators configs
[2021-03-08 19:01:36.823509000][I][Irohad]: [Init] => transaction batch parser
[2021-03-08 19:01:36.823739000][I][Irohad]: [Init] => validators
[2021-03-08 19:01:36.854718000][I][Irohad/Validators/Chain]: validate chain...
[2021-03-08 19:01:36.857246000][I][Irohad/Storage/Storage/MutableStorageImpl]: Applying block: height 1, hash 9debdb1a70db2cede2222427b849f6bf7ab20845da7c3db1837c0df25ec1c61a
[2021-03-08 19:01:36.878233000][I][Irohad]: [Init] => crypto provider
[2021-03-08 19:01:36.879281000][I][Irohad]: [Init] => factories
.........
```

```
>>> curl http://127.0.0.1:8080/metrics
# HELP exposer_transferred_bytes_total Transferred bytes to metrics services
# TYPE exposer_transferred_bytes_total counter
exposer_transferred_bytes_total 0.000000
# HELP exposer_scrapes_total Number of times metrics were scraped
# TYPE exposer_scrapes_total counter
exposer_scrapes_total 0.000000
# HELP exposer_request_latencies Latencies of serving scrape requests, in microseconds
# TYPE exposer_request_latencies summary
exposer_request_latencies_count 0
exposer_request_latencies_sum 0.000000
exposer_request_latencies{quantile="0.500000"} Nan
exposer_request_latencies{quantile="0.900000"} Nan
exposer_request_latencies{quantile="0.990000"} Nan
# HELP observed_packets_total Number of observed packets
# TYPE observed_packets_total counter
observed_packets_total{direction="tx",protocol="udp"} 1.000000
observed_packets_total{direction="rx",protocol="udp"} 1.000000
observed_packets_total{direction="tx",protocol="tcp"} 4.000000
observed_packets_total{direction="rx",protocol="tcp"} 3.000000
# HELP http_requests_total Number of HTTP requests
# TYPE http_requests_total counter
http_requests_total{method="GET"} 2.000000
http_requests_total{method="HEAD"} 2.000000
http_requests_total{method="POST"} 2.000000
http_requests_total{method="PUT"} 1.000000
```

### CODE REVIEW

File: .clang-format [hunk 1]
```diff
@@ -8,3 +8,4 @@ DerivePointerAlignment: false
 PointerAlignment: Right
 AllowShortFunctionsOnASingleLine: Empty
 AllowShortIfStatementsOnASingleLine: false
+TabWidth: 2
```
Comment (kamilsa) [inline .clang-format:11]: why?

File: irohad/main/iroha_conf_loader.cpp [hunk 2]
```diff
@@ -768,7 +769,7 @@ iroha::expected::Result<IrohadConfig, std::string> parse_iroha_config(
       return std::make_optional(std::move(doc));
     }};
 
-    JsonDeserializerImpl parser(common_objects_factory, doc, std::move(log));
+    JsonDeserializerImpl parser(common_objects_factory, doc, std::move(log));   //why move log?
```
Comment (kamilsa) [inline irohad/main/iroha_conf_loader.cpp:772]: log is copied into this function anyway. Nothing bad to move log I guess

File: irohad/maintenance/metrics.cpp [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+#include <regex>
+#include <optional>
+
+auto maintenance_metrics_init(std::string const& listen_addr)
+->std::optional<std::shared_ptr<prometheus::Registry>>
+{
+  using namespace prometheus;
+
+  std::regex full_matcher("^(([0-9]|[1-9][0-9]|1[0-9]{2}|2[0-4][0-9]|25[0-5])\\.){3}([0-9]|[1-9][0-9]|1[0-9]{2}|2[0-4][0-9]|25[0-5]):[0-9]+$");
+  std::regex port_matcher("^:?([0-9]{1,5})$");
```
Comment (kamilsa) [inline irohad/maintenance/metrics.cpp:28]: Make const static?

File: example/config.sample [hunk 4]
```diff
@@ -9,6 +9,7 @@
   "mst_enable" : false,
   "mst_expiration_time" : 1440,
   "max_rounds_delay": 3000,
-  "stale_stream_max_rounds": 2
+  "stale_stream_max_rounds": 2,
+  "metrics": "127.0.0.1:8080"
```
[the remark this answers is not in the document]
Comment (kamilsa) [reply example/config.sample:13]: ah, I see it could be either host:port or just port. Then let's keep it

**Review by kamilsa — COMMENTED** [review]
Please also add information about metrics endpoint to documentation here:
https://github.com/hyperledger/iroha/blob/master/docs/source/configure/index.rst

File: irohad/maintenance/metrics.cpp (hunk 3 above)
Comment (kuvaldini) [reply irohad/maintenance/metrics.cpp:28]: done

File: example/config.sample (hunk 4 above)
[the remark this answers is not in the document]
Comment (kuvaldini) [reply example/config.sample:13]: Yes.

File: .clang-format (hunk 1 above)
Comment (kuvaldini) [reply .clang-format:11]: removed

File: .clang-format [hunk 5]
```diff
@@ -7,4 +7,4 @@ BinPackParameters: false
 DerivePointerAlignment: false
 PointerAlignment: Right
 AllowShortFunctionsOnASingleLine: Empty
-AllowShortIfStatementsOnASingleLine: false
+AllowShortIfStatementsOnASingleLine: false
```
Comment (kamilsa) [inline .clang-format:10]: Please add empty line at the end

File: irohad/maintenance/metrics.cpp (hunk 3 above)
Comment (iceseer) [reply irohad/maintenance/metrics.cpp]: https://github.com/hyperledger/iroha/blob/b77adcb49e8f7a43ac7492400e308593027d96e6/shared_model/validators/field_validator.cpp#L82

File: irohad/maintenance/metrics.hpp [hunk 6]
```diff
@@ -0,0 +1,13 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#pragma once
```
Comment (iceseer) [inline irohad/maintenance/metrics.hpp:6]: The other code consistency:
``` c++
#ifndef IROHA_MAINTENANCE_METRICS_HPP
#define IROHA_MAINTENANCE_METRICS_HPP
{...}
#endif//IROHA_MAINTENANCE_METRICS_HPP

**Review by iceseer — CHANGES_REQUESTED** [review]

File: irohad/maintenance/metrics.cpp [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+#include <chrono>
+#include <cstdlib>
+#include <memory>
+#include <string>
+#include <thread>
+#include <future>
+#include <regex>
+
+auto maintenance_metrics_init(std::string const& listen_addr)
+->std::shared_ptr<prometheus::Registry>
```
Comment (iceseer) [inline irohad/maintenance/metrics.cpp:22]: You don't use compile-time return type calculation.
``` c++
std::shared_ptr<prometheus::Registry>
maintenance_metrics_init(std::string const& listen_addr)
```

**Review by iceseer — CHANGES_REQUESTED** [review]

**Review by kamilsa — APPROVED** [review]

File: irohad/maintenance/metrics.cpp (hunk 7 above)
Comment (kuvaldini) [reply irohad/maintenance/metrics.cpp:22]: Here I use only new syntax. But I changed it for consistancy

File: irohad/maintenance/metrics.hpp [hunk 8]
```diff
@@ -3,11 +3,14 @@
...
 #include <string>
 #include <memory>
 #include <prometheus/registry.h>
 
-auto maintenance_metrics_init(std::string const& listen_addr)
-->std::shared_ptr<prometheus::Registry>;
+std::shared_ptr<prometheus::Registry>
+    maintenance_metrics_init(std::string const& listen_addr);
+
+#endif //IROHA_MAINTENANCE_METRICS_HPP
```
Comment (kamilsa) [inline irohad/maintenance/metrics.hpp:16]: No empty line at the end

**Review by iceseer — APPROVED** [review]

**Pull request merged**